text/cognito-construct-library/working-backwards-readme.md

```diff
@@ -0,0 +1,300 @@
+  }
+});
+```
+
+> Internal Note: Implemented via UserPool-AdminCreateUserConfig, temp password via UserPool-Policies,
+> verifyContactMethods via UserPool-AutoVerifiedAttributes.
+
+### Sign In
+
+These are the various ways a user of your app can sign in. There are 4 options available with the enum `SignInType`:
+
+* USERNAME: Allow signing in using the one time immutable user name that the user chose at the time of sign up.
+* PREFERRED\_USERNAME: Allow signing in with an alternate user name that the user can change at any time. However, this
+  is not available if the USERNAME option is not chosen.
+* EMAIL: Allow signing in using the email address that is associated with the account.
+* PHONE\_NUMBER: Allow signing in using the phone number that is associated with the account.
+
+*Defaults*: USERNAME.
+
+Code sample:
+
+```ts
+new UserPool(this, 'myuserpool', {
+  // ...
+  // ...
+  signInType: [ SignInType.USERNAME, SignInType.EMAIL ],
+});
+```
+
+> Internal Note: Implemented via UserPool-UsernameAttributes and -AliasAttributes
+
+### Attributes
+
+These are the set of attributes you want to collect and store with each user in your user pool. Cognito provides a set
+of standard attributes that are available all user pools. Users are allowed to select any of these standard attributes
+to be required. Users will not be able to sign up without specifying the attributes that are marked as required. Besides
+these, additional attributes can be further defined, known as custom attributes.
+
+Custom attributes cannot be marked as required.
+
+[Go here](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html) for more info.
+
+Standard attributes are available via the `StandardAttrs` enum.
+
+*Note*: By default, the standard attributes 'email' and/or 'phone\_number' will automatically be marked required if
+they are one of the verified contact methods. See [Sign up](#sign-up) for details on verified contact methods.
+
+Custom attributes can be specified via the `stringAttr` and `numberAttr` methods, depending on whether the attribute
+type is either a string or a number. Constraints can be defined on both string and number types, with length constraint
+on the former and range constraint on the latter.
+
+Additionally, two properties `mutable` and `adminOnly` properties can be set for each custom attribute. The former
+specifies that the property can be modified by the user while the latter specifies that it can only be modified by the
+app's administrator and not by the user (using their access token).
+
+*Defaults*:
+* No standard attributes are marked required.
+* For all custom attributes, mutable is true and adminOnly is false.
+
+Code sample:
+
+```ts
+new UserPool(this, 'myuserpool', {
+  // ...
+  // ...
+  attributes: {
+    required: [ StandardAttrs.address, StandardAttrs.name ],
+    custom: [
+      stringAttr({ name: 'myappid', minLen: 5, maxLen: 15 }),
```

Comment:
I am assuming "name" must be unique, which lends itself to a map instead of an array:

```ts
attributes: {
  myappid: ...
  callingcode: { required: true }
}